docs/developer/benchmark.ipynb

```diff
@@ -0,0 +1,587 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "id": "3756777c",
+   "metadata": {},
+   "source": [
+    "# Benchmark\n",
+    "\n",
+    "ESS is built for various types of experiments with multiple instruments using different techniques.\n",
+    "\n",
+    "Benchmarking was needed to make sure the data reduction frameworks can handle the streaming data in real time.\n",
+    "\n",
+    "We will monitor the two types of computing costs, ``time`` and ``space(memory)`` of workflows."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "8be4d6e5",
+   "metadata": {},
+   "source": [
+    "## Running Benchmarks\n",
+    "\n",
+    "Benchmark tests and related tools are collected in ``tests/benchmarks``\n",
+    "and ``tests/prototypes`` of the [repository](https://github.com/scipp/beamlime).\n",
+    "\n",
+    "These tools and prototypes are implemented with dependency injection modules of ``beamlime``.\n",
+    "\n",
+    "### Benchmark Session\n",
+    "\n",
+    "``BenchmarkSession`` is the key helper for benchmark tests.\n",
+    "\n",
+    "It is a bridge object that glues \n",
+    "``report: BenchmarkReport``, ``runner: BenchmarkRunner`` and ``file_manager: BenchmarkFileManager``.\n",
+    "\n",
+    "``runner`` is expected to be customized often for various ways of tests.\n",
+    "``report`` and ``file_manager`` are not expected to be replaced as often.\n",
+    "\n",
+    "You can build a benchmark session object from the ``Factory``."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "3aaa96a4",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import sys, os\n",
+    "sys.path.insert(0, os.path.abspath('../../'))  # To use ``tests`` as a package."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "db5d1f7c",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from rich.pretty import Pretty\n",
+    "from tests.benchmarks.runner import create_benchmark_runner_factory, BenchmarkSession\n",
+    "\n",
+    "benchmark_factory = create_benchmark_runner_factory()\n",
+    "session = benchmark_factory[BenchmarkSession]\n",
+    "Pretty(session, max_depth=1)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "0d75c9c3",
+   "metadata": {},
+   "source": [
+    "Here is the simple use case of ``BenchmarkSession.run``."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "bc9eaaa0",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def test_func(x: float) -> float:\n",
+    "    from time import sleep\n",
+    "    sleep(x)\n",
+    "\n",
+    "    return x\n",
+    "\n",
+    "session.run(test_func, 0.1)\n",
+    "Pretty(session.report.measurements)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "7b63a66a",
+   "metadata": {},
+   "source": [
+    "The ``BenchmarkSession.run`` passes all arguments to ``runner`` and append its result into the ``report``.\n",
```

Comment:
I can't see `runner` in the code, what are you referring to? Same for `report`.

Reply:
I rephrased them to be consistent with the code blocks... sorry...!